### PR TITLE
fix(nodejs): fix infinite loop when package link from `package-lock.json` file is broken

### LINKED ISSUE #6854: bug(npm): runtime: out of memory
### Discussed in https://github.com/aquasecurity/trivy/discussions/6853

<div type='discussions-op-text'>

<sup>Originally posted by **gobardhan** June  4, 2024</sup>
### Description

For a npm project with following configuration of workspace inside package.json Trivy not able to perform the scan for any scanners.
`"workspaces": [
    "./"
  ]`

After debugging it I found that it's due the below code inside package-lock.json
`"node_modules/@ffdev/bryntum-components": {
      "resolved": "",
      "link": true
    },`
It's started working fine if I remove the above part of code from package-lock.json.

### Desired Behavior

Trivy should able to scan the repository with such kind of configuration for package.json & package-lock.json without throwing any error.

### Actual Behavior

Trvivy scan throw's `fatal error: runtime: out of memory`. Command `trivy fs .` 
And running trivy in debug mode hang-up my system and in last killed by system. Command `trvy fs . -d`
Screenshot from 2024-06-04 16-04-46

### Reproduction Steps

```bash
1. For any npm project inside the package.json file define workspace like ` "workspaces": ["./"] `
2. Do npm install
3. Execute trivy scanning by command `trivy fs .` OR `trivy fs . -d`
```

### Target

Filesystem

### Scanner

None

### Output Format

None

### Mode

Standalone

### Debug Output

```bash
trivy fs .
2024/06/04 15:58:56 INFO Loaded file_path=trivy.yaml
2024-06-04T15:58:56+05:30       INFO    Vulnerability scanning is enabled
2024-06-04T15:58:56+05:30       INFO    [npm] To collect the license information of packages, "npm install" needs to be performed beforehand dir="node_modules"
fatal error: runtime: out of memory

runtime stack:
runtime.throw({0x8423caf?, 0x2037cb?})
        /opt/hostedtoolcache/go/1.22.3/x64/src/runtime/panic.go:1023 +0x5c fp=0xc000613e10 sp=0xc000613de0 pc=0x43e67c
runtime.sysMapOS(0xc279000000, 0x1cacc00000)
        /opt/hostedtoolcache/go/1.22.3/x64/src/runtime/mem_linux.go:167 +0x11b fp=0xc000613e50 sp=0xc000613e10 pc=0x41c5bb
runtime.sysMap(0xc279000000, 0x1cacc00000, 0xd5ecaa8?)
.
.
.
```

### Operating System

Ubuntu 22.04.4 LTS

### Version

```bash
Version: 0.51.2
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-06-04 06:11:34.940587192 +0000 UTC
  NextUpdate: 2024-06-04 12:11:34.940586981 +0000 UTC
  DownloadedAt: 2024-06-04 08:58:30.997724006 +0000 UTC
Check Bundle:
  Digest: sha256:274208f3b043e73d2f1e5d9baed30d77c33bb164b8f2e580020e9c5931c71df4
  DownloadedAt: 2024-06-03 09:58:55.819249061 +0000 UTC
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)</div>

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
This PR fixes 2 cases:
1. When `Packages[x].resolved` is empty for link - we don't need to resolve this link.
2. Don't overwrite `packages` map when resolving links to avoid cases when we updated package is parsed.

## Related issues
- Close #6854

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/dependency/parser/nodejs/npm/parse.go [hunk 1]
```diff
@@ -208,7 +216,8 @@ func (p *Parser) resolveLinks(packages map[string]Package) {
 	}
 
 	workspaces := rootPkg.Workspaces
-	for pkgPath, pkg := range packages {
+	// Clone packages to avoid cases when we check already updated packages
```
Comment (knqyf263) [inline pkg/dependency/parser/nodejs/npm/parse.go:219]: Could you elaborate on that?

Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/npm/parse.go:219]: We update `packages` map inside of `packages` loop.
https://github.com/aquasecurity/trivy/blob/a2d1871387e511040e7eb8368d62c266b0197763/pkg/dependency/parser/nodejs/npm/parse.go#L232-L235
this is bad practice.

When I tested broken link with empty `resolved` field - this loop would endlessly update  `packages`, check them and  we get  `out of memory` error.

Comment (knqyf263) [reply pkg/dependency/parser/nodejs/npm/parse.go:219]: Can we change the comment to something like "Changing the map during the map iteration causes unexpected behavior" for clarity?

Comment (DmitriyLewen) [reply pkg/dependency/parser/nodejs/npm/parse.go]: Yeah, it make sense.
Updated in https://github.com/aquasecurity/trivy/pull/6858/commits/b8e52c499fa94ebf59191fdcbf8b935e6584402a

**Review by knqyf263 — APPROVED** [review]

**Comment by DmitriyLewen** [comment]
@aqua-bot backport release/v0.52

**Pull request merged**